augur/io/metadata.py

```diff
@@ -88,3 +97,228 @@ def read_metadata(metadata_file, id_columns=("strain", "name"), chunk_size=None)
+                if record_id is None:
+                    raise AugurError(f"The provided id column {id_column!r} does not exist in the metadata.")
+
+                if record_id in seen_ids:
+                    duplicate_ids.add(record_id)
+                else:
+                    seen_ids.add(record_id)
+
+                yield record
+
+    if duplicate_reporting is not DataErrorMethod.SILENT and duplicate_ids:
+        duplicates_message = f"The following records are duplicated in {metadata!r}:\n" + "\n".join(sorted(duplicate_ids))
+
+        if duplicate_reporting is DataErrorMethod.WARN:
+            print_err(f"WARNING: {duplicates_message}")
+        elif duplicate_reporting is DataErrorMethod.ERROR:
+            raise AugurError(duplicates_message)
+        else:
+            raise AugurError(f"Encountered unhandled duplicate reporting method: {duplicate_reporting!r}")
+
+
+def read_metadata_with_sequences(metadata, fasta, seq_id_column, seq_field='sequence',
+    unmatched_reporting=DataErrorMethod.ERROR, duplicate_reporting=DataErrorMethod.ERROR):
+    """
+    Read rows from *metadata* file and yield each row as a single dict that has
+    been updated with their corresponding sequence from the *fasta* file.
+    Matches the metadata record with sequences using the sequence id provided
+    in the *seq_id_column*. To ensure that the sequences can be matched with
+    the metadata, the FASTA headers should only contain the sequence id.
+
+    Will report unmatched records if requested via *unmatched_reporting* after
+    the generator has been exhausted.
+
+    Will report duplicate records if requested via *duplicate_reporting* after
+    the generator has been exhausted.
+
+    Reads the *fasta* file with `pyfastx.Fasta`, which creates an index for
+    the file to allow random access of sequences via the sequence id.
+    See pyfastx docs for more details:
+    https://pyfastx.readthedocs.io/en/latest/usage.html#fasta
+
+    Parameters
+    ----------
+    metadata: str
+        Path to a CSV or TSV metadata file
+
+    fasta: str
+        Path to a plain or gzipped FASTA file
+
+    seq_id_column: str
+        The column in the metadata file that contains the sequence id for
+        matching sequences
+
+    seq_field: str, optional
+        The field name to use for the sequence in the updated record
+
+    unmatched_reporting: DataErrorMethod, optional
+        How should unmatched records be reported
+
+    duplicate_reporting: DataErrorMethod, optional
+        How should duplicate records be reported
+
+    Yields
+    ------
+    dict
+        The parsed metadata record with the sequence
+    """
+    sequences = pyfastx.Fasta(fasta)
+    sequence_ids = set(sequences.keys())
+
+    # Used for determining unmatched records
+    processed_sequence_ids = set()
+    unmatched_metadata_ids = set()
+
+    # Used for determining duplicate records
+    seen_metadata_ids = set()
+    duplicate_metadata_ids = set()
```

Comment:
`processed_` and `seen_` represent the same concept here. Can just one be used for both variable names?

Reply:
There's a variable named `seen_sequence_ids` further down for checking duplicates within the sequences:

https://github.com/nextstrain/augur/blob/682d3d3c36a593293aa9462fd94551e13d5701ad/augur/io/metadata.py#L286

I'll update both to be `processed_` here.

Reply:
Fixed up in 0b637d298c9b90f3f76f82f00af4e806bb63b960